src/collector.py

```diff
@@ -0,0 +1,451 @@
+
+    MAX = "max_in_millis"
+    AVG = "avg_in_millis"
+
+
+class RequestsCount(Enum):
+    """Possible requests count types for the metrics"""
+
+    DISCONNECTS = "disconnects"
+    TOTAL = "total"
+
+
+class DashBoardsCollector(Collector):
+    """OpenSearch Dashboards Collector"""
+
+    def __init__(self, config: Config) -> None:
+        super().__init__()
+        self.config = config
+
+    def collect(self) -> Generator[Metric, None, None]:
+        """Collect all the metrics from /api/status to be exported.
+
+        Yields:
+            Generator[Metric]: the internal status metrics of status
+        """
+        api_metrics = collect_api_status(self.config)
+        api_up = GaugeMetricFamily(
+            name=f"{METRICS_PREFIX}up",
+            documentation="Whether the data source is reachable (1 for up, 0 for down)",
+            value=1 if api_metrics else 0,
+        )
+        yield api_up
+
+        for metric in self.metrics(api_metrics):
+            description, value = metric
+            if value:
+                yield value
+            else:
+                logger.error("It was not possible to get the metric: %s", description)
+
+    def metrics(self, api_metrics: dict) -> list[tuple[str, Optional[Metric]]]:
+        """Get the OpenSearch dashboard prometheus metrics.
+
+        Args:
+            api_metrics (dict): Metrics from the OpenSearch Dashboards API
+
+        Returns:
+            list[tuple[str, Optional[Metric]]]: Prometheus Gauge metrics of the dashboards
+        """
+        if api_metrics:
+            return [
+                ("status", _get_overall_status_metric(api_metrics)),
+                ("current_connections", _get_current_connections_metric(api_metrics)),
+                ("up_time", _get_up_time_metric(api_metrics)),
+                ("event_loop_delay", _get_event_loop_delay_metric(api_metrics)),
+                ("heap_total", _get_heap(api_metrics, Heap.TOTAL)),
+                ("heap_used", _get_heap(api_metrics, Heap.USED)),
+                ("heap_size", _get_heap(api_metrics, Heap.SIZE)),
+                ("re_set_size", _get_resident_set_size(api_metrics)),
+                ("load_1m", _get_load(api_metrics, Load.ONE_M)),
+                ("load_5m", _get_load(api_metrics, Load.FIVE_M)),
+                ("load_15m", _get_load(api_metrics, Load.FIFTEEN_M)),
+                ("os_mem_total", _get_os_mem(api_metrics, Memory.TOTAL)),
+                ("os_mem_free", _get_os_mem(api_metrics, Memory.FREE)),
+                ("os_mem_used", _get_os_mem(api_metrics, Memory.USED)),
+                ("resp_time_avg", _get_resp_time(api_metrics, Response.AVG)),
+                ("resp_time_max", _get_resp_time(api_metrics, Response.MAX)),
+                ("req_disconnects", _get_req(api_metrics, RequestsCount.DISCONNECTS)),
+                ("req_total", _get_req(api_metrics, RequestsCount.TOTAL)),
+            ] + [("statuses", status) for status in _get_statuses_metrics(api_metrics)]
```

Comment:
I'm a little concerned re: this last bit.  If we have multiple ```("statuses": X)``` tuples (as the code implies is possible), what would the data generated actually look like?  Would it make sense, despite the "statuses" key being used multiple times?

Reply:
In the end we will have multiple `opensearch_dashboards_statuses` Gauge metrics and on it's labels it's where we can identify by the `id`.

In another words, with the API example in the code we would have:
("statuses", GaugeMetricFamily), ("statuses", GaugeMetricFamily)

The first one about `core:opensearch@2.17.0` and the second one about `plugin:securityDashboards@2.17.0`

Then I could just simply create an alert rule that check opensearch_dashboards_statuses == 2 to trigger alerts if one of those plugins are in red state.

Does it make sense?